bugtool/cmd/root.go

```diff
@@ -0,0 +1,237 @@
+	src := filepath.Join(defaults.RuntimePath, defaults.StateDir)
+	dst := filepath.Join(dbgDir, defaults.StateDir)
+	if err := copyDir(src, dst); err != nil {
+		log.WithError(err).Warn("could not copy state directory")
+	}
+}
+
+func serveStaticFiles(debugDirectory string) {
+	fs := http.FileServer(http.Dir(debugDirectory))
+	addr := fmt.Sprintf(":%d", port)
+
+	http.Handle("/", fs)
+	log.Printf("Serving files at http://localhost%s", addr)
+	if err := http.ListenAndServe(addr, nil); err != nil {
+		log.WithError(err).Fatal("could not start server")
+	}
+}
+
+func createDir(dbgDir string, newDir string) string {
+	confDir := filepath.Join(dbgDir, newDir)
+	if err := os.Mkdir(confDir, defaults.RuntimePathRights); err != nil {
+		log.WithError(err).Warn("failed to create %s info directory", newDir)
+		return dbgDir
+	}
+	return confDir
+}
+
+func copyKernelConfig(confDir string) {
+	kernel, _ := execCommand("uname", "-r")
+	kernel = strings.TrimSpace(kernel)
+	locations := []struct {
+		Src string
+		Dst string
+	}{
+		{"/proc/config", fmt.Sprintf("%s/kernel-config", confDir)},
+		{"/proc/config.gz", fmt.Sprintf("%s/kernel-config.gz", confDir)},
+		{fmt.Sprintf("/boot/config-%s", kernel), fmt.Sprintf("%s/kernel-config-%s", confDir, kernel)},
+	}
+
+	for _, location := range locations {
+		if _, err := os.Stat(location.Src); os.IsNotExist(err) {
+			continue
+		}
+		err := copyFile(location.Src, location.Dst)
+		if err != nil {
+			log.WithError(err).Warn("could not copy kernel config")
+		}
+	}
+}
+
+func copySystemCmds(cmdDir string) {
+	// Not expecting all of the commands to be available.
+	sources := []string{
+		// Host and misc
+		"hostname", "ip a", "ip r", "ip link", "iptables-save", "uname -a",
+		"dig", "netstat", "pidstat", "arp", "top -b -n 1",
+		"uptime", "dmesg", "bpftool map show", "bpftool prog show",
+		"iptables -S", "ip6tables -S", "iptables -L -v", "ps",
+		"cat /proc/sys/net/core/bpf_jit_enable", "cat /proc/kallsyms",
+		// Versions
+		"docker version",
+		"kubectl version",
+		// Try to find some useful logs
+		"cat /var/log/upstart/docker.log",
+		"cat /var/log/docker.log", "cat /var/log/daemon.log",
+		"journalctl -u cilium*", "cat /var/log/messages", "journalctl -u kubelet",
+		"kubectl -n kube-system get pods",
+		"kubectl get pods,svc --all-namespaces",
+		"kubectl get version",
+	}
+
+	// Append ethtool links
+	if links, err := netlink.LinkList(); err == nil {
+		for _, link := range links {
+			// query current settings
+			sources = append(sources, fmt.Sprintf("ethtool %s", link.Attrs().Name))
+			// query for driver information
+			sources = append(sources, fmt.Sprintf("ethtool -i %s", link.Attrs().Name))
+		}
+	}
+
+	execs := make(chan bool)
+	runInBackground := func(dir, cmd string) {
+		writeCmdToFile(dir, cmd)
+		execs <- true
+	}
+	for _, cmd := range sources {
+		go runInBackground(cmdDir, cmd)
+	}
+	// Wait for the commands to complete
+	for i := 0; i < len(sources); i++ {
+		<-execs
+	}
```

Comment:
Is the intention to force the commands to run one by one? If so, why not just spawn one goroutine that does the `for _, cmd := range sources` loop but executes the commands directly?
If the goal is to later allow more than one to run at once, this is a good way to do it (although in that case, I would recommend `make(chan bool, 1)` to make that more clear).

Reply:
Will change it to the later.

Reply:
Oh, in fact, I misread this. Because we write to the channel after we run the command, all we were/are doing was making the main routine wait for all the spawned goroutines to complete. What you have will do this, or you can use a WaitGroup (https://golang.org/pkg/sync/#WaitGroup) that does pretty must the same thing.

If the goal was to have only 2 (or whatever) commands run at once it's a little more complicated but:
```golang
	numRoutinesAtOnce := 2
	semaphore := make(chan bool, numRoutinesAtOnce)
	for i := 0; i < numRoutinesAtOnce; i++ {
		// This will not block because the channel is buffered and we can write to it numRoutinesAtOnce before the write blocks
		semaphore <- true
	}

	for _, cmd := range commands {
		wg := syncWaitGroup()
		// tell the waitgroup it needs to track another goroutine
		wg.Add(1)

		// start a subroutine to run our command
		go func() {
			// once we exit this goroutine completely, signal the original that we are done
			defer wg.Done()

			// This will wait until an entry in this channel is available to read. We started with numRoutinesAtOnce in there (from above)
			<-semaphore
			// When we are done we return the thing we took from the semaphore, so another goroutine can get it
			defer func() { semaphore <- true }()
			cmd()
		}()
	}
	// wait for all the spawned goroutines to finish up.
	wg.Wait()
```